### PR TITLE
feat(core): add Mobile Payment (Direct Carrier Billing) as a payment method

### PR DESCRIPTION
**Pull request by Sakilmostak**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Following are changes added in the pr:
- New payment method: Mobile Payment
- New payment method type: Direct Carrier Billing
- It is a network based billing payment method where OTP(one time password) is required for the payment confirmation. 
- New next action type - Collect OTP
- This next action gives an URL in the response through which you collect OTP for the transaction
- New connector added - Digital Virgo
- For the connector integration: Auth, Sync and Error handling have been implemented

### Additional Changes

- [x] This PR modifies the API contract
In request, following payment method  is added:
```
"payment_method": "mobile_payment",
    "payment_method_type": "direct_carrier_billing",
    "payment_method_data": {
        "mobile_payment": {
            "direct_carrier_billing": {
                "msisdn": "48506000000",
                "client_uid": "02iacdYXGI9CnyJddfsoN8c7"
            }
        }
    }
```
In response we should receive next action as:
```
"next_action": {
        "type": "collect_otp",
        "collect_otp_url": "{otp_url}"
    }
```

- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

- Create a MCA with Digital Virgo (add the following payment method): 
```
"payment_methods_enabled": [
        {
            "payment_method": "mobile_payment",
            "payment_method_types": [
                {
                    "payment_method_type": "direct_carrier_billing",
                    "payment_experience": "redirect_to_url",
                    "minimum_amount": 1,
                    "maximum_amount": 68607706,
                    "recurring_enabled": true,
                    "installment_payment_enabled": true
                }
            ]
        }
    ],
```

- Create a payment request: 
```
{
    "amount": 110,
    "currency": "USD",
    "confirm": true,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "customer_id": "StripeCustomer",
    "email": "abcdef123@gmail.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+65",
    "description": "Its my first payment request",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
    "setup_future_usage": "on_session",
    "browser_info": {
        "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36",
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "language": "nl-NL",
        "color_depth": 24,
        "screen_height": 723,
        "screen_width": 1536,
        "time_zone": 0,
        "java_enabled": true,
        "java_script_enabled": true,
        "ip_address": "127.0.0.1"
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "John",
            "last_name": "Doe"
        }
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    },
    "payment_method": "mobile_payment",
    "payment_method_type": "direct_carrier_billing",
    "payment_method_data": {
        "mobile_payment": {
            "direct_carrier_billing": {
                "msisdn": "48506000000",
                "client_uid": "02iacdYXGI9CnyJddfsoN8c7"
            }
        }
    },
    "surcharge_details": {
        "surcharge_amount": 50
    },
    "order_details": [
        {
            "product_name": "FLOWBIRD",
            "quantity": 1,
            "amount": 110
        }
    ],
    "billing": {
        "address": {
            "line1": "8th block",
            "line2": "8th block",
            "line3": "8th block",
            "city": "Bengaluru",
            "state": "Karnataka",
            "zip": "560095",
            "country": "IN",
            "first_name": "Sakil",
            "last_name": "Mostak"
        }
    }
}
```

- The response will of following type:
```
{
    "payment_id": "pay_WGRM7PpvBRSlSDZniNEk",
    "merchant_id": "merchant_1728629559",
    "status": "requires_customer_action",
    "amount": 110,
    "net_amount": 160,
    "amount_capturable": 160,
    "amount_received": null,
    "connector": "digitalvirgo",
    "client_secret": "pay_WGRM7PpvBRSlSDZniNEk_secret_0UfF4oF4boWQ8achbdYj",
    "created": "2024-10-11T06:54:06.173Z",
    "currency": "USD",
    "customer_id": "StripeCustomer",
    "customer": {
        "id": "StripeCustomer",
        "name": "John Doe",
        "email": "abcdef123@gmail.com",
        "phone": "999999999",
        "phone_country_code": "+65"
    },
    "description": "Its my first payment request",
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": "on_session",
    "off_session": null,
    "capture_on": null,
    "capture_method": "automatic",
    "payment_method": "mobile_payment",
    "payment_method_data": {
        "mobile_payment": {
            "direct_carrier_billing": {
                "msisdn": "48506000000",
                "client_uid": "02iacdYXGI9CnyJddfsoN8c7"
            }
        },
        "billing": null
    },
    "payment_token": null,
    "shipping": {
        "address": {
            "city": "San Fransico",
            "country": "US",
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "zip": "94122",
            "state": "California",
            "first_name": "John",
            "last_name": "Doe"
        },
        "phone": null,
        "email": null
    },
    "billing": {
        "address": {
            "city": "Bengaluru",
            "country": "IN",
            "line1": "8th block",
            "line2": "8th block",
            "line3": "8th block",
            "zip": "560095",
            "state": "Karnataka",
            "first_name": "Sakil",
            "last_name": "Mostak"
        },
        "phone": null,
        "email": null
    },
    "order_details": [
        {
            "brand": null,
            "amount": 110,
            "category": null,
            "quantity": 1,
            "product_id": null,
            "product_name": "FLOWBIRD",
            "product_type": null,
            "sub_category": null,
            "product_img_link": null,
            "product_tax_code": null,
            "requires_shipping": null
        }
    ],
    "email": "abcdef123@gmail.com",
    "name": "John Doe",
    "phone": "999999999",
    "return_url": "https://google.com/",
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "next_action": {
        "type": "collect_otp",
        "collect_otp_url": "http://localhost:8080/payments/redirect/pay_WGRM7PpvBRSlSDZniNEk/merchant_1728629559/pay_WGRM7PpvBRSlSDZniNEk_1"
    },
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": "direct_carrier_billing",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": {
        "customer_id": "StripeCustomer",
        "created_at": 1728629646,
        "expires": 1728633246,
        "secret": "epk_c4eec95dcda24d39b6c4ebfe13b0e479"
    },
    "manual_retry_allowed": null,
    "connector_transaction_id": "f24f3bb69d34438d",
    "frm_message": null,
    "metadata": {
        "udf1": "value1",
        "login_date": "2019-09-10T10:11:12Z",
        "new_customer": "true"
    },
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": null,
    "payment_link": null,
    "profile_id": "pro_mPEx1vKwPqr1By68pRQH",
    "surcharge_details": {
        "surcharge_amount": 50,
        "tax_amount": 0
    },
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_JZI7HRbAAG2hnqd436mZ",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-10-11T07:09:06.173Z",
    "fingerprint": null,
    "browser_info": {
        "language": "nl-NL",
        "time_zone": 0,
        "ip_address": "127.0.0.1",
        "user_agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.110 Safari/537.36",
        "color_depth": 24,
        "java_enabled": true,
        "screen_width": 1536,
        "accept_header": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8",
        "screen_height": 723,
        "java_script_enabled": true
    },
    "payment_method_id": null,
    "payment_method_status": null,
    "updated": "2024-10-11T06:54:06.689Z",
    "charges": null,
    "frm_metadata": null,
    "merchant_order_reference_id": null,
    "order_tax_amount": null,
    "connector_mandate_id": null
}
```

- Redirect to the url and fill the otp to confirm the payment.
- Do a retrieve call
- The status should be succeeded in response of retrieve call

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/hyperswitch_connectors/src/connectors/digitalvirgo.rs [hunk 1]
```diff
@@ -168,9 +234,9 @@ impl ConnectorIntegration<Authorize, PaymentsAuthorizeData, PaymentsResponseData
     fn get_url(
         &self,
         _req: &PaymentsAuthorizeRouterData,
-        _connectors: &Connectors,
+        connectors: &Connectors,
     ) -> CustomResult<String, errors::ConnectorError> {
-        Err(errors::ConnectorError::NotImplemented("get_url method".to_string()).into())
+        Ok(format!("{}{}", self.base_url(connectors), "/payment"))
```
Comment (srujanchikke) [inline crates/hyperswitch_connectors/src/connectors/digitalvirgo.rs:239]: Please make the same changes in all get_url functions.

```suggestion
        Ok(format!("{}/payment", self.base_url(connectors)))
```

File: crates/hyperswitch_connectors/src/connectors/digitalvirgo.rs [hunk 2]
```diff
@@ -355,13 +445,13 @@ impl ConnectorIntegration<Capture, PaymentsCaptureData, PaymentsResponseData> fo
...
-        data: &PaymentsCaptureRouterData,
+        data: &PaymentsCompleteAuthorizeRouterData,
         event_builder: Option<&mut ConnectorEvent>,
         res: Response,
-    ) -> CustomResult<PaymentsCaptureRouterData, errors::ConnectorError> {
+    ) -> CustomResult<PaymentsCompleteAuthorizeRouterData, errors::ConnectorError> {
         let response: digitalvirgo::DigitalvirgoPaymentsResponse = res
             .response
-            .parse_struct("Digitalvirgo PaymentsCaptureResponse")
+            .parse_struct("Digitalvirgo PaymentsAuthorizeResponse")
```
Comment (srujanchikke) [inline crates/hyperswitch_connectors/src/connectors/digitalvirgo.rs:454]:
```suggestion
            .parse_struct("DigitalvirgoPaymentsResponse")
```

**Review by srujanchikke — CHANGES_REQUESTED** [review]

**Review by srujanchikke — APPROVED** [review]
Reviewed connector code!

**Review by deepanshu-iiitu — APPROVED** [review]

File: crates/router/src/core/payments/transformers.rs [hunk 3]
```diff
@@ -907,6 +907,12 @@ where
 
         let next_action_voucher = voucher_next_steps_check(payment_attempt.clone())?;
 
+        let next_action_mobile_payment = mobile_payment_next_steps_check(
+            base_url,
+            payment_attempt.clone(),
+            payment_intent.clone(),
```
Comment (Chethan-rao) [inline crates/router/src/core/payments/transformers.rs:913]: clone not required here, you can pass the reference

File: crates/euclid/src/frontend/dir.rs [hunk 4]
```diff
@@ -278,6 +278,13 @@ pub enum DirKeyKind {
         props(Category = "Payment Method Types")
     )]
     OpenBankingType,
+    #[serde(rename = "mobile_payment")]
+    #[strum(
+        serialize = "mobile_payment",
+        detailed_message = "Supported types of mobile payment payment method",
```
Comment (Chethan-rao) [inline crates/euclid/src/frontend/dir.rs:284]:
```suggestion
        detailed_message = "Supported types of mobile payment method",
```

**Review by Chethan-rao — CHANGES_REQUESTED** [review]

**Review by Chethan-rao — APPROVED** [review]
Routing and core changes LGTM

**Review by deepanshu-iiitu — APPROVED** [review]
Approving on behalf of connectors.

**Review by srujanchikke — DISMISSED** [review]

**Comment by Narayanbhat166** [comment]
> This next action gives an URL in the response through which you collect OTP for the transaction

Is this like a redirection? If so then we can send the redirect url right?

**Comment by Sakilmostak** [comment]
> Is this like a redirection? If so then we can send the redirect url right?

Yes, this is a redirection but generally redirection is either done by connector or is used to invoke their sdk. Here, we are collecting the OTP with our own hosted site to confirm the payment.

**Review by Chethan-rao — DISMISSED** [review]

File: crates/hyperswitch_connectors/src/connectors/digitalvirgo/transformers.rs [hunk 5]
```diff
@@ -54,63 +54,80 @@ impl TryFrom<&DigitalvirgoRouterData<&PaymentsAuthorizeRouterData>>
...
-            }
+            PaymentMethodData::MobilePayment(mobile_payment_data) => match mobile_payment_data {
+                MobilePaymentData::DirectCarrierBilling { msisdn, client_uid } => {
+                    let order_details = item.router_data.request.get_order_details()?;
+                    let product_name = order_details
+                        .first()
+                        .map(|order| order.product_name.to_owned())
+                        .ok_or(errors::ConnectorError::MissingRequiredField {
+                            field_name: "product_name",
+                        })?;
```
Comment (srujanchikke) [inline crates/hyperswitch_connectors/src/connectors/digitalvirgo/transformers.rs:65]: Should required fields to be added in payment_connector_required_fields.rs ?

**Review by srujanchikke — DISMISSED** [review]

**Review by srujanchikke — APPROVED** [review]

File: crates/api_models/src/payments.rs [hunk 6]
```diff
@@ -3225,6 +3242,20 @@ pub enum OpenBankingData {
...
     OpenBankingPIS {},
 }
+
+#[derive(Eq, PartialEq, Clone, Debug, serde::Deserialize, serde::Serialize, ToSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum MobilePaymentData {
+    DirectCarrierBilling {
+        /// The phone number of the user
+        #[schema(value_type = String, example = "1234567890")]
+        msisdn: String,
```
Comment (sai-harsha-vardhan) [inline crates/api_models/src/payments.rs:3252]: This should be a Secret value

File: crates/api_models/src/payments.rs [hunk 7]
```diff
@@ -3225,6 +3242,20 @@ pub enum OpenBankingData {
...
+        /// Unique user id
+        #[schema(value_type = Option<String>, example = "02iacdYXGI9CnyJdoN8c7")]
+        client_uid: Option<String>,
```
Comment (sai-harsha-vardhan) [inline crates/api_models/src/payments.rs:3255]: Is this sensitive data?

**Review by sai-harsha-vardhan — APPROVED** [review]

**Review by Chethan-rao — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Comment by Gnanasundari24** [comment]
@Sakilmostak Can you add the cypress testcases for this

**Comment by Sakilmostak** [comment]
The flow is incomplete for the new payment method. The subsequent PR as stated by this issue (https://github.com/juspay/hyperswitch-cloud/issues/7441) should be able to complete the flow. The cypress test will be raised along with this PR

**Pull request merged**